### DISCUSSION TITLE
Auth -- access_token in local storage?

### DISCUSSION THREAD

**Discussion by rwbrockhoff** [supabase/supabase#21824] [Questions]
We have a React frontend with a Node/express backend. When I set up the Auth in React though, I noticed it stored the access_token, refresh_token, and user info in local storage. Services like Kinde (and security recommendations in general) prefer using httpOnly cookies to store this information. It seems like we could potentially manage auth completely server-side where there is the option to store session/httpOnly cookie, but I can't find a way to do this client-side. 

It seems like Supabase has great security. Yet localStorage is being used? Can someone help me understand why this approach is safe since it goes against what other services are doing? And how is it a good idea to store the user info in local storage too? Totally open to using this solution if there isn't the security risk that I'm imagining.

**Comment by akabutnicer** [comment]
Use indexeddb

**Comment by akabutnicer** [comment]
A global indexeddb reassigning to localStorage should work

**Reply by rwbrockhoff** [reply]
I'm more so asking if there's a means within the API or a preferred Supabase approach to using a secure cookie. Or a security explanation on why local storage is being used. But I appreciate your input!

**Reply by akabutnicer** [reply]
Well, I'm not so sure about that. The only options you really have is to use indexeddb or just gather the information needed for the account then unassign localStorage

**Reply by akabutnicer** [reply]
Although they[d still be able to check the application in devtools

**Reply by akabutnicer** [reply]
I can provide code for the indexedb if you would like

**Comment by akabutnicer** [comment]
And also, I have the exact same project set up, I just unassign localStorage since at least then its slightly less exposed

**Reply by rwbrockhoff** [reply]
You could set the token as a httpOnly secure cookie on the backend after an auth state change in the frontend and delete the localStorage on a 200 success response so that way it's even more secure. But I'm not sure how that plays into their token rotation/refresh. And if you delete everything from local storage, I'm not sure the API would know where to look for refreshing things. Is Supabase still updating tokens when you put them in the indexedb?

**Reply by akabutnicer** [reply]
If you proxy methods correctly, the static methods it calls will be transferred to your indexeddb. You could also do it with cookies; for example:
```javascript
localStorage = new Proxy({}, {
    get: function(target, prop, receiver) {
        return unescape((document.cookie.match(prop + '=([^;].+?)(;|$)') || [])[1] || '');
    }
});
Object.defineProperty(localStorage, 'setItem', [
    writable: false,
    value: function(key, value) {
        return (document.cookie = `${name}=${value}`)
    }
});
Object.defineProperty(localStorage, 'getItem', {
    writable: false,
    value: function(name) {
        return unescape((document.cookie.match(name + '=([^;].+?)(;|$)') || [])[1] || null);
    }
})
```

**Comment by krispekla** [comment]
Since this is very important topic, I think it needs more discussion because as I am seeing sb is auto setting access token info in local storage. And as it seems for security best practices http only cookie is recommended.

**Comment by NordlingDev** [comment]
Seems dangerous to me that localStorage is the default storage in Supabase considering that an access token has personal information stored in it (i.e. email address). LocalStorage never gets cleared automatically like safe httpOnly cookies as they expire eventually and prevent token hijacking by XSS. LocalStorage is extremely vulnerably in this regard.

**Comment by Felipe-Tomazetti** [comment]
No return from Supabase regarding this issue?

**Reply by zaczacariah** [reply]
This seems quite important - I wish they would shed some light

**Reply by krispekla** [reply]
For me also this is deal breaker.

**Reply by zaczacariah** [reply]
Having had a little more look...the only real concern is that if your token expires in 1 hour then within that hour someone could impersonate you if XXS were used. 

I did find this: https://supabase.com/docs/guides/auth/server-side/advanced-guide 
which seems to outline how to use session cookies instead of local storage. I haven't tried using this with the UI library but i imagine it will be problematic.

**Comment by skiunke** [comment]
Stumbled upon this as well, using Supabase with Flutter and the package
